### PR TITLE
BTRX - Project Review - Fundings

### PR DESCRIPTION
**Pull request by LeoForconesi**
## Changes
- Implement funding review, edition and compare in project review.
- Fix elements comparison order in Multi select component.
- Correct console warning for uncontrolled variable.

## Notes
- You will find a couple of `todo` flags, these indicate some improvements coming in the next PR for saving code and complexity.

---

- [x] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Get a thumb from @rushtong
- [ ] **Submitter**: Merge to develop using github's "Squash and Merge" feature
- [ ] **Submitter**: Delete branch after merge

### CODE REVIEW

File: src/main/webapp/components/Fundings.js [hunk 1]
```diff
@@ -111,11 +184,14 @@ export const Fundings = hh(class Fundings extends Component {
                       index: idx,
                       name: "source",
                       options: fundingOptions,
-                      value: this.props.fundings[idx].source,
+                      value: this.props.edit ? rd.future.source : rd.source,
+                      currentValue: this.props.edit ? current[idx].current.source : rd.source,
                       onChange: this.handleFundingSelect,
-                      error: this.props.error && idx === 0,
+                      error: this.props.edit === true  && this.props.errorIndex !== null? this.props.error && this.props.errorIndex.includes(idx) : this.props.error && idx === 0,
```
Comment (rushtong) [inline src/main/webapp/components/Fundings.js]: To be 100% honest, these kinds of multi-conditional checks make my eyes cross. This seems like a good place to abstract the logic into a function that is easier to grok.

File: src/main/webapp/project/NewProjectGeneralData.js [hunk 2]
```diff
@@ -216,7 +216,7 @@ export const NewProjectGeneralData = hh(class NewProjectGeneralData extends Comp
             id: "inputStudyActivitiesDescription",
             name: "studyDescription",
             label: "Describe Broad study activities* ",
-            moreInfo: "(briefly, in 1-2 paragraphs, with attention to wheter or not protected health information will be accessed, future data sharing plans, and commercial or academic sample/data sources. For commercially purchased products, please cite product URL.)",
+            moreInfo: "(briefly, in 1-2 paragraphs, with attention to whether or not protected health information will be accessed, future data sharing plans, and commercial or academic sample/data sources. For commercially purchased products, please cite product URL.)",
```
Comment (rushtong) [inline src/main/webapp/project/NewProjectGeneralData.js]: Nice catch

File: src/main/webapp/projectReview/ProjectReview.js [hunk 3]
```diff
@@ -45,14 +46,19 @@ class ProjectReview extends Component {
...
+        }],
         requestor: {
           displayName: '',
           emailAddress: ''
         }
       },
       disableApproveButton: false,
       reviewSuggestion: false,
+      futureCopy: {
+      },
```
Comment (rushtong) [inline src/main/webapp/projectReview/ProjectReview.js]: nit: can be a single line: `futureCopy: {},`

File: src/main/webapp/projectReview/ProjectReview.js [hunk 4]
```diff
@@ -330,19 +354,38 @@ class ProjectReview extends Component {
...
       });
     }
     return fundingList;
   }
 
   compareObj(obj1, obj2) {
-    return JSON.stringify(this.state[obj1]) === JSON.stringify(this.state[obj2]);
+    let form1 = JSON.parse(JSON.stringify(this.state[obj1]));
+    let form2 = JSON.parse(JSON.stringify(this.state[obj2]));
+    if ( form1 !== undefined && form2 !== undefined && form1.fundings !== undefined && form2.fundings !== undefined) {
```
Comment (rushtong) [inline src/main/webapp/projectReview/ProjectReview.js]: Could you do something like this instead? (https://www.w3schools.com/jsref/jsref_every.asp)
```
[form1, form2, form1.fundings, form2.fundings].every(function(el){ return el !== undefined })
```

Comment (rushtong) [reply src/main/webapp/projectReview/ProjectReview.js]: Alternatively, we could use lodash.get which has a form of null-safe traversal: https://lodash.com/docs/#get

Comment (rushtong) [reply src/main/webapp/projectReview/ProjectReview.js]: I admit to being spoiled by Groovy's null-safe handling - it's one of the things that keeps me coming back to languages like it and Kotlin.

File: src/main/webapp/components/Fundings.js (hunk 1 above)
Comment (LeoForconesi) [reply src/main/webapp/components/Fundings.js]: This is now corrected, I must have overlooked it...

File: src/main/webapp/projectReview/ProjectReview.js (hunk 4 above)
Comment (LeoForconesi) [reply src/main/webapp/projectReview/ProjectReview.js]: I really liked lodash approach, so I've imported it to the project. Thanks for the suggestion! now we can count on this library.

**Pull request merged**